<b>In last class we learned about why we use rnn on sequential data rather than ann because rnns have memory feature and they can remember the sementic meaning between any sequential dataset

<b>In this class we will learn how Rnn makes prediction on unseen data also known as <mark> Forward Propogation

When we send data in RNN The format of data is :(timesteps,input_features)

<mark><strong><b> Data Feeding in RNN

# <span style="color:#1B4F72;">How Data is Sent to an RNN?</span>

An RNN processes **sequential data one step at a time** while carrying information from previous time steps through its **hidden state**.

---

## <span style="color:#2980B9;">Example Dataset</span>

Suppose we have these reviews:

```text
1. movie was good  →  1
2. movie was bad   →  0
3. movie was not good →  0
```

Here:

```text
1 → Positive
0 → Negative
```

---

## <span style="color:#8E44AD;">Vocabulary</span>

First, we create a vocabulary containing all unique words:

```text
movie → 0
was   → 1
good  → 2
bad   → 3
not   → 4
```

So:

```text
Vocabulary Size = 5
```

---

## <span style="color:#16A085;">Convert Words into Numbers</span>

Each word can be represented using **One-Hot Encoding**.

```text
movie → [1,0,0,0,0]
was   → [0,1,0,0,0]
good  → [0,0,1,0,0]
bad   → [0,0,0,1,0]
not   → [0,0,0,0,1]
```

For:

```text
movie was good
```

we get:

```text
[
 [1,0,0,0,0],   ← movie
 [0,1,0,0,0],   ← was
 [0,0,1,0,0]    ← good
]
```

Therefore, its shape is:

```text
(3, 5)
```

> **3 = number of time steps**  
> **5 = number of input features / vocabulary size**

---

## <span style="color:#C0392B;">What is a Time Step?</span>

In an RNN, each word can be treated as one **time step**.

```text
movie → t₁
was   → t₂
good  → t₃
```

The words are **not sent all at once**.

They are processed sequentially:

```text
t₁              t₂              t₃
movie           was             good
  ↓              ↓               ↓
RNN ───────────→ RNN ──────────→ RNN
  ↓              ↓               ↓
h₁              h₂              h₃
```

At every step, the RNN receives:

```text
Current Input + Previous Hidden State
```

---

## <span style="color:#D35400;">How Information Flows</span>

At the first time step:

```text
movie
  ↓
RNN
  ↓
h₁
```

At the second time step:

```text
was + h₁
      ↓
     RNN
      ↓
     h₂
```

At the third time step:

```text
good + h₂
       ↓
      RNN
       ↓
      h₃
```

So the information flows like:

```text
Input₁ → RNN → h₁
                 ↓
Input₂ → RNN → h₂
                 ↓
Input₃ → RNN → h₃
```

> The **hidden state carries information from previous time steps**.

---

## <span style="color:#7D3C98;">For Sentiment Classification</span>

For a complete review:

```text
movie → was → good
  ↓      ↓      ↓
 RNN →  RNN →  RNN
  ↓      ↓      ↓
 h₁     h₂     h₃
               ↓
          Final Output
               ↓
          Sentiment = 1
```

The final hidden state contains information gathered from the sequence and can be used to make the final prediction.

```text
"movie was good"
        ↓
      RNN
        ↓
   Sentiment
        ↓
        1
```

---

## <span style="color:#2980B9;">Input Shape in PyTorch</span>

For an RNN, the input is generally represented as:

```text
(batch_size, sequence_length, input_size)
```

For example:

```text
(1, 3, 5)
```

means:

```text
1 → one review
3 → three time steps / words
5 → five input features
```

For multiple reviews:

```text
(batch_size, 3, 5)
```

Example:

```text
(3, 3, 5)
```

means:

```text
3 reviews
3 words per review
5 features per word
```

---

## <span style="color:#1B4F72;">Complete Flow</span>

```text
Raw Text
   ↓
Tokenization
   ↓
Vocabulary
   ↓
Numerical Representation
   ↓
Sequence
   ↓
(batch_size, seq_len, input_size)
   ↓
RNN processes one time step at a time
   ↓
Hidden State carries previous information
   ↓
Final Hidden State
   ↓
Output Layer
   ↓
Prediction
```

> **Core Idea:**  
> **RNN does not simply process all words independently. It processes the sequence step-by-step and carries information from previous steps through the hidden state.**

in Pytorch we send in this format : (sequence_length, batch_size, input_size)

<b>This is how we provide data to rnn to work

<b><mark> How RNN Works Intenally

# <span style="color:#1B4F72;">ANN vs RNN — Major Differences</span>

There are **2 major differences** between ANN and RNN.

---

## <span style="color:#2980B9;">Difference 1: How Data is Processed</span>

### **ANN**

In ANN, we generally provide the **complete input at once** to the neural network.

```text
Complete Input
      ↓
Forward Propagation
      ↓
Output
      ↓
Backpropagation
      ↓
Update Weights
```

### **RNN**

In RNN, **sequential data is processed step-by-step (one time step at a time)**.

```text
Input₁ → RNN → Hidden State₁
                  ↓
Input₂ → RNN → Hidden State₂
                  ↓
Input₃ → RNN → Hidden State₃
                  ↓
                Output
```

The important difference is that the RNN passes the **hidden state from the previous time step** to the next time step.

> **ANN → Processes the input as a whole**  
> **RNN → Processes the sequence step-by-step + carries previous information**

## <span style="color:#8E44AD;">Difference 2: Data Flow & Hidden State</span>

### **ANN**

ANN does **not maintain a hidden state or memory** of previous inputs.

```text
Input → Neural Network → Output
```

Each input is processed independently.

### **RNN**

RNN maintains a **hidden state** that carries information from the previous time step.

```text
Input₁ → RNN → Hidden State₁
                  ↓
                  Input₂ → RNN → Hidden State₂
                                    ↓
                                    Input₃ → RNN → gHidden State₃
                                    ```

                                    The **previous hidden state is passed back into the RNN** along with the current input.

                                    > **ANN → No memory of previous inputs**  
                                    > **RNN → Hidden state carries information from previous time steps**

In [ ]:
# code implementation

import torch
import torch.nn as nn

In [ ]:
rnn = nn.RNN(
    input_size=5,
    hidden_size=3,
    batch_first=True
)

dense = nn.Linear(3, 1)

In [ ]:
print(rnn)
print(dense)

RNN(5, 3, batch_first=True)
Linear(in_features=3, out_features=1, bias=True)


In [ ]:
print(rnn.weight_ih_l0)
print(rnn.weight_hh_l0)

print(rnn.bias_ih_l0)
print(rnn.bias_hh_l0)

Parameter containing:
tensor([[ 0.0392, -0.4541,  0.4431, -0.2525,  0.3940],
        [ 0.1590,  0.1349, -0.5183,  0.3884, -0.1615],
        [-0.4190,  0.0672, -0.3807, -0.0209, -0.1153]], requires_grad=True)
Parameter containing:
tensor([[ 0.1655, -0.4445, -0.4782],
        [-0.0088, -0.0284,  0.2306],
        [-0.2211,  0.4669, -0.5718]], requires_grad=True)
Parameter containing:
tensor([-0.4536, -0.3311,  0.1126], requires_grad=True)
Parameter containing:
tensor([-0.1188,  0.4625, -0.4489], requires_grad=True)


<mark> Forward Propagation In Rnn

<b>we learned how we provide data to rnn and now we learn Forward Propagation inside rnn it means when we give inputs to rnn how does rnn calculates the output

Concpet used While FP in rnn : Unfolding thorough time

<b> <mark> Full Picture how rnn process data and find output (detailed )

# 🧠 RNN Forward Propagation — Sentiment Analysis Example

<div align="center">

# 🔄 RNN Forward Propagation
### From Words → Vectors → Hidden States → Final Sentiment

</div>

---

## 🎯 1. Objective

We will understand **RNN forward propagation** using a very small
**sentiment-analysis dataset with only 4 sentences**.

Our task:

> Given a movie review, predict whether it is **Positive (1)** or **Negative (0)**.

---

## 📊 2. Sample Dataset

| Review | Sentiment |
|---|---:|
| `good movie` | 1 |
| `bad movie` | 0 |
| `very good` | 1 |
| `very bad` | 0 |

For understanding the RNN, we will take only **one review**:

> **"good movie"**

So our sequence contains **2 time steps**:

```text
t = 1 → good
t = 2 → movie


---

🧩 3. Convert Words into Vectors

An RNN cannot directly understand words.

So suppose we have already converted our words into simple vectors:

good  → [1, 0, 1, 0, 0]
movie → [0, 1, 0, 1, 0]

Therefore:

\[
X_1 =
\begin{bmatrix}
1\\
0\\
1\\
0\\
0
\end{bmatrix}
\]

\[
X_2 =
\begin{bmatrix}
0\\
1\\
0\\
1\\
0
\end{bmatrix}
\]

Important dimensions

Input vector size = 5

Therefore:

\[
\boxed{input\_size = 5}
\]

We have:

Number of words = 2

Therefore:

\[
\boxed{sequence\ length = 2}
\]

For this example, suppose:

batch_size = 1

Hence the input shape is:

\[
\boxed{(batch\_size,\ sequence\_length,\ input\_size)}
\]

\[
\boxed{(1,\ 2,\ 5)}
\]


---

🧠 4. RNN Architecture

Suppose our RNN is:

rnn = nn.RNN(
    input_size=5,
    hidden_size=3,
    batch_first=True
)

dense = nn.Linear(3, 1)

Therefore:

Input size  = 5
Hidden size = 3
Output size = 1

The RNN will create a hidden vector containing 3 values.


---

🔑 5. RNN Formula

The basic RNN hidden-state equation is:

\[
\boxed{
h_t = f(W_{ih}x_t + b_{ih} + W_{hh}h_{t-1} + b_{hh})
}
\]

Where:

Symbol  Meaning

(x_t)   Current input
(h_t)   Current hidden state
(h_{t-1})   Previous hidden state
(W_{ih})    Input → Hidden weights
(W_{hh})    Hidden → Hidden weights
(b_{ih})    Input → Hidden bias
(b_{hh})    Hidden → Hidden bias
(f) Activation function


Usually, for a simple RNN:

\[
f = \tanh
\]

So:

\[
\boxed{
h_t =
\tanh(W_{ih}x_t+b_{ih}+W_{hh}h_{t-1}+b_{hh})
}
\]


---

⏱️ 6. Time Step 1 — Word = "good"

Our first word is:

\[
X_1 =
\begin{bmatrix}
1\\
0\\
1\\
0\\
0
\end{bmatrix}
\]

At the beginning, there is no previous hidden state.

Therefore we initialize:

\[
\boxed{
h_0 =
\begin{bmatrix}
0\\
0\\
0
\end{bmatrix}
}
\]

So:

Previous hidden state
        ↓
h₀ = [0, 0, 0]


---

⚙️ 7. Assume Some Weights

To make the calculation easy, let's use small illustrative weights.

Input → Hidden weights

\[
W_{ih} =
\begin{bmatrix}
0.2 & 0.1 & 0.3 & 0.1 & 0.2\\
0.1 & 0.3 & 0.2 & 0.2 & 0.1\\
0.3 & 0.2 & 0.1 & 0.1 & 0.2
\end{bmatrix}
\]

Shape:

\[
\boxed{W_{ih}=(3,5)}
\]

Why?

hidden_size × input_size
= 3 × 5


---

Hidden → Hidden weights

\[
W_{hh} =
\begin{bmatrix}
0.1 & 0.2 & 0.1\\
0.2 & 0.1 & 0.2\\
0.1 & 0.1 & 0.2
\end{bmatrix}
\]

Shape:

\[
\boxed{W_{hh}=(3,3)}
\]


---

Biases

For simplicity:

\[
b_{ih} =
\begin{bmatrix}
0\\
0\\
0
\end{bmatrix}
\]

\[
b_{hh} =
\begin{bmatrix}
0\\
0\\
0
\end{bmatrix}
\]


---

🧮 8. Calculate Time Step 1

Formula:

\[
h_1 =
\tanh(W_{ih}X_1+b_{ih}+W_{hh}h_0+b_{hh})
\]

Since:

\[
h_0=[0,0,0]^T
\]

we get:

\[
W_{hh}h_0=0
\]

Therefore:

\[
h_1=\tanh(W_{ih}X_1)
\]

Now:

\[
W_{ih}X_1
=
\begin{bmatrix}
0.2 & 0.1 & 0.3 & 0.1 & 0.2\\
0.1 & 0.3 & 0.2 & 0.2 & 0.1\\
0.3 & 0.2 & 0.1 & 0.1 & 0.2
\end{bmatrix}
\begin{bmatrix}
1\\
0\\
1\\
0\\
0
\end{bmatrix}
\]

Calculate each value:

\[
0.2(1)+0.1(0)+0.3(1)+0.1(0)+0.2(0)
=0.5
\]

\[
0.1(1)+0.3(0)+0.2(1)+0.2(0)+0.1(0)
=0.3
\]

\[
0.3(1)+0.2(0)+0.1(1)+0.1(0)+0.2(0)
=0.4
\]

Therefore:

\[
W_{ih}X_1=
\begin{bmatrix}
0.5\\
0.3\\
0.4
\end{bmatrix}
\]

Apply tanh:

\[
h_1=
\tanh
\begin{bmatrix}
0.5\\
0.3\\
0.4
\end{bmatrix}
\]

Approximately:

\[
\boxed{
h_1\approx
\begin{bmatrix}
0.462\\
0.291\\
0.380
\end{bmatrix}
}
\]


---

🔄 9. Time Step 2 — Word = "movie"

Now we send the second word.

\[
X_2=
\begin{bmatrix}
0\\
1\\
0\\
1\\
0
\end{bmatrix}
\]

But this time something important happens.

We have a previous hidden state:

\[
h_1=
\begin{bmatrix}
0.462\\
0.291\\
0.380
\end{bmatrix}
\]

Therefore the RNN receives TWO pieces of information:

┌──────────────┐
X₂ = "movie" ───►│              │
                 │     RNN      │───► h₂
h₁ ─────────────►│              │
                 └──────────────┘

This is the fundamental idea of an RNN:

> Current input + previous hidden state → current hidden state




---

🧮 10. Calculate Time Step 2

Formula:

\[
\boxed{
h_2=
\tanh(W_{ih}X_2+b_{ih}+W_{hh}h_1+b_{hh})
}
\]

Let's calculate the two major components separately.


---

Part A — Current Word

\[
W_{ih}X_2
\]

\[
=
\begin{bmatrix}
0.2 & 0.1 & 0.3 & 0.1 & 0.2\\
0.1 & 0.3 & 0.2 & 0.2 & 0.1\\
0.3 & 0.2 & 0.1 & 0.1 & 0.2
\end{bmatrix}
\begin{bmatrix}
0\\
1\\
0\\
1\\
0
\end{bmatrix}
\]

First neuron:

\[
0.1+0.1=0.2
\]

Second neuron:

\[
0.3+0.2=0.5
\]

Third neuron:

\[
0.2+0.1=0.3
\]

Therefore:

\[
W_{ih}X_2=
\begin{bmatrix}
0.2\\
0.5\\
0.3
\end{bmatrix}
\]


---

🧮 11. Previous Information

Now calculate:

\[
W_{hh}h_1
\]

\[
=
\begin{bmatrix}
0.1 & 0.2 & 0.1\\
0.2 & 0.1 & 0.2\\
0.1 & 0.1 & 0.2
\end{bmatrix}
\begin{bmatrix}
0.462\\
0.291\\
0.380
\end{bmatrix}
\]

Approximately:

\[
W_{hh}h_1
\approx
\begin{bmatrix}
0.143\\
0.225\\
0.151
\end{bmatrix}
\]


---

➕ 12. Combine Current + Previous Information

Now:

\[
W_{ih}X_2 + W_{hh}h_1
\]

\[
=
\begin{bmatrix}
0.2\\
0.5\\
0.3
\end{bmatrix}
+
\begin{bmatrix}
0.143\\
0.225\\
0.151
\end{bmatrix}
\]

\[
=
\begin{bmatrix}
0.343\\
0.725\\
0.451
\end{bmatrix}
\]

Now apply tanh:

\[
h_2=
\tanh
\begin{bmatrix}
0.343\\
0.725\\
0.451
\end{bmatrix}
\]

Approximately:

\[
\boxed{
h_2\approx
\begin{bmatrix}
0.330\\
0.620\\
0.423
\end{bmatrix}
}
\]


---

🧠 13. What Just Happened?

This is the most important part.

At:

⏱️ T = 1

good
 ↓
X₁
 ↓
RNN
 ↓
h₁

There was no previous information, so:

\[
h_0=0
\]


---

⏱️ T = 2

movie ──────► X₂
                │
                ▼
              ┌─────┐
h₁ ──────────►│ RNN │
              └─────┘
                │
                ▼
               h₂

Now:

\[
\boxed{
h_2=f(X_2,h_1)
}
\]

So (h_2) contains information from:

"good"
   +
"movie"

This is how the RNN processes sequence information.


---

🔁 14. Unfolding Through Time

The RNN can be visualized as:

┌─────────┐
X₁ ─────────►│   RNN   │
             └────┬────┘
                  │
                 h₁
                  │
                  ▼
             ┌─────────┐
X₂ ─────────►│   RNN   │
             └────┬────┘
                  │
                 h₂
                  │
                  ▼
             ┌─────────┐
X₃ ─────────►│   RNN   │
             └────┬────┘
                  │
                 h₃

This representation is called:

\[
\boxed{\text{Unfolding Through Time}}
\]

⚠️ Important

These are NOT different RNN layers.

They are the same RNN cell reused at different time steps.

Therefore:

\[
\boxed{\text{Same weights are reused at every time step}}
\]

This is called:

\[
\boxed{\text{Parameter Sharing / Weight Sharing}}
\]


---

📐 15. Complete RNN Forward Pass

For a sequence:

\[
X_1,X_2,\ldots,X_T
\]

the RNN performs:

\[
\boxed{
h_1=f(W_{ih}X_1+W_{hh}h_0+b)
}
\]

\[
\boxed{
h_2=f(W_{ih}X_2+W_{hh}h_1+b)
}
\]

\[
\boxed{
h_3=f(W_{ih}X_3+W_{hh}h_2+b)
}
\]

and so on:

\[
\boxed{
h_t=f(W_{ih}X_t+W_{hh}h_{t-1}+b)
}
\]


---

🎯 16. From Hidden State to Sentiment

After processing the complete sentence, we can use the last hidden state:

\[
h_2
\]

to predict sentiment.

Our dense layer is:

dense = nn.Linear(3, 1)

Therefore:

\[
\boxed{
y_{raw}=W_yh_2+b_y
}
\]

Here:

h₂ → 3 values
     ↓
Linear Layer
     ↓
1 value

Suppose:

\[
W_y=
\begin{bmatrix}
0.5 & 0.8 & 0.2
\end{bmatrix}
\]

and:

\[
b_y=0.1
\]

Then:

\[
y_{raw}
=
0.5(0.330)+0.8(0.620)+0.2(0.423)+0.1
\]

\[
y_{raw}\approx0.706
\]

This raw value is called a logit.


---

🔢 17. Convert Logit to Probability

For binary sentiment classification, we can use the sigmoid function:

\[
\boxed{
\sigma(z)=\frac{1}{1+e^{-z}}
}
\]

Therefore:

\[
P(\text{Positive})
=
\sigma(0.706)
\]

\[
\boxed{
P(\text{Positive})\approx0.67
}
\]

So:

Positive probability ≈ 67%
Negative probability ≈ 33%

Using a threshold of 0.5:

\[
P(\text{Positive})>0.5
\]

Therefore:

\[
\boxed{\text{Prediction = Positive}}
\]


---

🧠 18. Complete Flow

<div align="center">"good movie"
      │
      ▼
Tokenization
      │
      ▼
Word → Vector
      │
      ├───────────────┐
      ▼               │
   X₁ = good          │
      │               │
      ▼               │
    RNN                │
      │               │
      ▼               │
     h₁ ──────────────┤
                      │
                      ▼
   X₂ = movie ──────► RNN
                      │
                      ▼
                     h₂
                      │
                      ▼
                  Linear Layer
                      │
                      ▼
                    Logit
                      │
                      ▼
                   Sigmoid
                      │
                      ▼
              Positive / Negative

</div>
---

📦 19. Dimensions at Every Step

For our example:

batch_size   = 1
sequence_len = 2
input_size   = 5
hidden_size  = 3
output_size  = 1

Input:

\[
X=(1,2,5)
\]

At each time step:

\[
X_t=(1,5)
\]

Hidden state:

\[
h_t=(1,3)
\]

Final hidden state:

\[
h_2=(1,3)
\]

Linear layer:

\[
(1,3)\rightarrow(1,1)
\]

Final output:

\[
\boxed{(1,1)}
\]


---

🔥 20. The Core Concept to Remember

The entire RNN forward propagation can be reduced to this:

\[
\boxed{
\text{Current Input}
+
\text{Previous Hidden State}
\rightarrow
\text{Current Hidden State}
}
\]

Mathematically:

\[
\boxed{
h_t=
\tanh
\left(
W_{ih}x_t+
W_{hh}h_{t-1}+
b
\right)
}
\]

For our sentence:

\[
\boxed{
h_1=f(X_1,h_0)
}
\]

\[
\boxed{
h_2=f(X_2,h_1)
}
\]

Therefore:

\[
\boxed{
h_2\text{ contains information from }X_1\text{ and }X_2
}
\]

And that is the fundamental reason we use an RNN for sequential data. 🔄🧠


---

⭐ One-Line Mental Model

> RNN = Read the current word + remember information from previous words → create a new memory (hidden state).



Word₁ → h₁
         ↓
Word₂ + h₁ → h₂
              ↓
Word₃ + h₂ → h₃
              ↓
             ...
              ↓
       Final Hidden State
              ↓
          Prediction

<b> <mark> Simple words : using dot products+ output of previous time stamps and also the weights of the nn